lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -1611,13 +1616,36 @@ def _set_unit_ip(self, _):
         event is actually needed.
         """
         for relation in self._charm.model.relations[self._relation_name]:
-            relation.data[self._charm.unit]["prometheus_scrape_unit_address"] = str(
-                self._charm.model.get_binding(relation).network.bind_address
-            )
+            unit_ip = str(self._charm.model.get_binding(relation).network.bind_address)
+
+            if not self._is_valid_unit_address(unit_ip):
+                # relation data will be updated later when a valid address becomes available
+                if "prometheus_scrape_unit_address" in relation.data[self._charm.unit]:
+                    del relation.data[self._charm.unit]["prometheus_scrape_unit_address"]
+                    del relation.data[self._charm.unit]["prometheus_scrape_unit_name"]
```

Comment:
Technically this could throw a key error.
Maybe
```suggestion
                with contextlib.suppress(KeyError):
                    del relation.data[self._charm.unit]["prometheus_scrape_unit_address"]
                    del relation.data[self._charm.unit]["prometheus_scrape_unit_name"]
```

Reply:
FWIW please note changes suggested like this risk the introduction of a bug because they do not import `contextlib` into this module at the same time. I will make this change locally and make another push.